Implement WALReader

adsharma opened the pull request:
This could be useful to external replication tools to read WAL
records similar to how wal2json (Postgres) and binlog (MySQL) work.

Translation to externally consumable format is not included.

adsharma committed:
Add ParseInfo types for most WAL records

adsharma committed:
In deserialize_only mode return parsed records

Previously we deserialized the records and ignored them. Now we
return a unique_ptr<>

adsharma committed:
Implement WALReader

This could be useful to external replication tools to read WAL
records similar to how wal2json (Postgres) and binlog (MySQL) work.

Translation to externally consumable format is not included.

adsharma commented:
More context here: https://github.com/duckdb/duckdb/discussions/17200

Tishj commented on src/include/duckdb/parser/parsed_data/sequence_value_info.hpp:
```diff
@@ -0,0 +1,39 @@
+
+public:
+	unique_ptr<SequenceValueInfo> Copy();
+	string ToString() const;
```
This doesn't look like it's defined anywhere?

adsharma replied:
Resolved in the first PR on the stack

Tishj commented on src/include/duckdb/parser/parsed_data/table_data_info.hpp:
```diff
@@ -0,0 +1,118 @@
+	string name;
+
+public:
+	unique_ptr<TableDataInfo> Copy();
```
The same comment above applies to this and the other ParseInfo derived classes that are added

Tishj commented on src/storage/wal_replay.cpp:
```diff
@@ -112,43 +115,49 @@ class WriteAheadLogDeserializer {
-	void ReplayUpdate();
-	void ReplayCheckpoint();
+protected:
+	std::unique_ptr<ParseInfo> ReplayEntry(WALType wal_type);
```
`unique_ptr`, or `duckdb::unique_ptr`

Tishj commented on src/storage/wal_replay.cpp:
```diff
@@ -360,37 +344,39 @@ void WriteAheadLogDeserializer::ReplayEntry(WALType entry_type) {
+std::unique_ptr<ParseInfo> WriteAheadLogDeserializer::ReplayVersion() {
 	state.wal_version = deserializer.ReadProperty<idx_t>(101, "version");
+	auto info = make_uniq<ParseInfo>(ParseInfoType::VERSION_INFO);
+	return info;
```
I don't understand the benefit of returning an empty ParseInfo here, don't we want to add the `state.wal_version` to this?

adsharma replied:
Returning `nullptr` causes the `WALReader` to terminate the loop prematurely. It's interpreted as "no more WAL records" left. I didn't return the version because `WALReader` can get it from `ReplayState`.

Tishj commented on src/storage/wal_replay.cpp:
```diff
@@ -763,58 +764,96 @@ void WriteAheadLogDeserializer::ReplayRowGroupData() {
+	}
+	// WAL file version, so Open() has the reader positioned at the expected offset.
+	//
+	// Simplified logic from ReplayVersion(). Note that we can't call the deserializer
```
This feels weird, and looking at the code, I'm not sure this is currently functioning correctly on `main` either ?

<https://github.com/duckdb/duckdb/pull/14785> recently increased the `WAL_VERSION_NUMBER`, which is what gets written by `WriteVersion`.

While `ReplayState` is created with `wal_version = 1` always, which makes this check here:
```c++
		if (state_p.wal_version == 1) {
			// old WAL versions do not have checksums
			return WriteAheadLogDeserializer(state_p, stream, deserialize_only);
		}
```
always trigger, making it so that the rest of the function body is never executed

@Mytherin

Tishj replied:
So to summarize, this is actually more correct than what happens anywhere else, and we should fix that.

adsharma commented:
Test failures are from the newly added tests:

* Windows: complains that WAL file is already open by unittest.exe and can't be opened again (even in read only mode?). Sounds like this is caused by how the WAL file is opened by the WAL writer. Would love to hear from someone more familiar with windows about potentially relaxing exclusivity.

* Linux Debug: I can't reproduce the failures. Retesting after remove some more pessimizing std::move() statements.

karannparikh commented on src/storage/wal_replay.cpp:
```diff
@@ -763,58 +764,96 @@ void WriteAheadLogDeserializer::ReplayRowGroupData() {
+	//
+	// An alternative is to call reader->Read<idx_t>(), ignore the return value and use
+	// duckdb::WAL_VERSION_NUMBER to set the wal_version.
+	auto wal_version = (reader->Read<idx_t>() >> 40) & 0xFF;
+	state = make_uniq<ReplayState>(database, *con.context);
+	state->wal_version = wal_version;
 }
```
I suggesting changing the code to the following, which then reads the first entry which fetches the correct WAL version and updates the state. 

```suggestion
    deserializer = WriteAheadLogDeserializer::Open(*state, *reader, true);
    _ = deserializer.DeserializeEntry();
}
```

adsharma replied:
```
	reader->Read<idx_t>();
	state = make_uniq<ReplayState>(database, *con.context);
	state->wal_version = 2;  // There is a named constant in another file, will need to move it to a header
```

is the only alternative that passes unit tests.

karannparikh reviewed (approved):
lgtm!

adsharma commented:
@Mytherin the included unit test passes on my Linux machine locally in both release/debug modes. Looking for suggestions on how to fix the CI failures.

adsharma commented:
@Tishj - yes I noticed them in the logs a few days ago. Looking for advice on how you'd like to see them tackled.

* Linux failures due to dereferencing nullptr. I can't reproduce it locally. My guess is that it's got something to differences between std::unique_ptr<> and duckdb::nullptr<> usage. 

* Windows failure. Don't have a windows box or knowledge of APIs to fix. My searches tell me that it should be possible to have the same file open twice on windows.

Tishj commented on test/wal_reader/wal_reader_test.cpp:
```diff
@@ -0,0 +1,298 @@
+			REQUIRE(empty_reader.Next() == nullptr);
+
+			// Test reading with invalid file handle
+			REQUIRE_THROWS_AS(WALReader(attached_db, nullptr), Exception);
```
I don't think you want to use this, in tests we run with "abort on internal exception" configuration, this causes an internal exception, which aborts.
I suggest to just remove this line.

adsharma commented:
Sending an update with remaining instance of `std::unique_ptr<>` removed and offending test line removed.
Windows failure is still TBD

Tishj commented:
> Sending an update with remaining instance of `std::unique_ptr<>` removed and offending test line removed. Windows failure is still TBD

The linux/mac file system locking behavior is different from windows, so I'm not sure, but it sounds like you're trying to open two database instance to the same physical file, at the same time.

adsharma commented:
@Tishj the windows test failure is likely related to [this line](https://github.com/duckdb/duckdb/blob/e0e4cc2cc80e59b13f792756529e36a0300365a4/src/common/local_file_system.cpp#L908) of code.

Would you be open to changing it to allow the same semantics as other platforms where opening a file in `FileLockType::WRITE_LOCK` will block other writers, but allow readers?

adsharma commented:
That didn't work. Here's the (easier to read) [documentation](https://ss64.com/nt/syntax-share-modes.html
) on windows sharing modes. Here's what could work:

WAL writer: GENERIC_WRITE, FILE_SHARING_READ 
WAL reader: GENERIC_READ, FILE_SHARE_WRITE

However, there is a mismatch between the semantics of `FileLockType` in the duckdb C++ API (which is written from the perspective of the current process) and window's `share_mode` (which is written from the perspective of the other process
you want to share with).

adsharma commented:
Using `FileLockType::READ_LOCK` breaks Linux:

```
[0/3] (0%): Test basic WAL reader functionality                                 Query failed with message: TransactionContext Error: Failed to commit: Could not set lock on file "duckdb_unittest_tempdir/60859/wal_reader_test.wal": Conflicting lock is held in PID 0. See also https://duckdb.org/docs/stable/connect/concurrency
```

Strangely `WRITE_LOCK` works. How is `WRITE_LOCK` weaker than `READ_LOCK` on Linux?

Testing without any `FileLockType` flags now.

adsharma commented:
I'm in the process of writing some tests, so we can discuss:

* What type of locks should WAL code use?
* Do other users of `BufferedFileWriter` need locks?
* How do we make the locks work on Windows/Linux/Mac with well understood semantics (#17418)?
  * I'll attach the tests to the discussion

adsharma commented:
Linux: this strace output explains why opening the WAL file with `FileLockType::READ_LOCK` fails:

```
openat(AT_FDCWD, "duckdb_unittest_tempdir/99693/wal_reader_test.wal", O_WRONLY|O_CREAT|O_APPEND|O_CLOEXEC, 0666) = 4                                
fstat(4, {st_mode=S_IFREG|0664, st_size=0, ...}) = 0       
fcntl(4, F_SETLK, {l_type=F_RDLCK, l_whence=SEEK_SET, l_start=0, l_len=0}) = -1 EBADF (Bad file descriptor)
```

man page says `O_WRONLY` is incompatible with `F_RDLCK`.

This PR is getting too busy. I'll move the discussion around a solution to #17418. Could be a separate PR that this one rebases on top of.

Mytherin commented:
Thanks for the PR!

Can we turn the WAL reader into a table function that is exposed to SQL instead? e.g.:

```sql
SELECT * FROM read_wal('file.duckdb.wal')
```

This could then return a number of records, e.g. in the following format:

```sql
record_type  VARCHAR
record_size BIGINT
record_info MAP(VARCHAR, VARCHAR)
record_data VARCHAR[][]
```

That could then look something like this:

```sql
CREATE TABLE tbl(col1 BIGINT, col2 BIGINT);
INSERT INTO tbl VALUES (1, 4), (2, 5), (3, 6);

┌─────────────┬─────────────┬──────────────────────────┬────────────────────────┐
│ record_type │ record_size │       record_info        │      record_data       │
│   varchar   │    int32    │  map(varchar, varchar)   │      varchar[][]       │
├─────────────┼─────────────┼──────────────────────────┼────────────────────────┤
│ insert      │     288     │ {schema=main, table=tbl} │ [[1, 2, 3], [4, 5, 6]] │
└─────────────┴─────────────┴──────────────────────────┴────────────────────────┘

```

That would then enable much easier writing of tests, and also make this a useful user-facing feature.

Mytherin commented:
> > you're trying to open two database instance to the same physical file, at the same time.
> 
> One database. The WAL writer has the WAL file open for writing. I've used a read transaction to prevent the WAL from being applied to storage, so it could be read without races. Then I use the WAL reader to open the WAL file a second time for reading. This is how other DBs do replication.
> 
> All of it works on Linux/mac. But on windows, I see this behavior.
> 
> If this is not workable on windows, another option is to copy the WAL file and then read the copy. This introduces additional I/O and code complexity. Would like to avoid if possible.

Instead of messing with file locks - you can create a WAL by disabling checkpointing on shutdown. See any of our WAL tests:

```sql
statement ok
PRAGMA disable_checkpoint_on_shutdown;

statement ok
SET wal_autocheckpoint='1TB';
```

You can then attach a database, perform some writes, and detach the database - this will leave behind a WAL that can be used without any locks attached to it.

adsharma commented:
> SELECT * FROM read_wal('file.duckdb.wal')

I'm doing the spiritual equivalent of this from python. Should be possible to make it a UDF.
The current design is to run the `WALReader` thread in the same process as follows:

```
    wal_replay_thread = WALReplayThread(db_path="test.db")
    wal_replay_thread.start()

    try:
        time.sleep(1)
        con = wal_replay_thread.local_con
        create_tables(con)
        write_synthetic_data(con)
        wreader = walreader.walreader.open(con)
        While True:
            entry = wreader.next()
            if entry is None:
                break
            print(entry)
            if isinstance(entry, walreader.TableDataInsertInfo):
                print(con.to_record_batch(entry.chunk))
            elif ...
        time.sleep(10)
```

The core algorithm is similar to [this](https://litestream.io/how-it-works/#the-shadow-wal). Most importantly, for live replication `walreader` takes a `db_path` as input and monitors all writes to the WAL. Uses the checkpointing + long-running read tx to control when the WAL gets written and applied.

`con.to_record_batch(entry.chunk)` was a trivial extension to the code in `python_udf.cpp`.

But the WAL records as they stand now make references to duckdb internal concepts such as `DataChunk`, visibility mask and `rowid`. They need to be translated to things that other processes can consume. `pyarrow::RecordBatch` sounded like a good solution.

Sometimes a single SQL statement results in multiple WAL records. In order to have a stable WAL API that an external reader can consume, sometimes we have to combine information from these records to get basic things like:

* What schema/table is getting modified?
* What cols are modified?
* Translate duckdb internal `rowid` to a primary key column

Still thinking through these topics. #17200 may be a good place for discussions that go beyond this PR.

Mytherin commented:
I don't really see any reason to involve pyarrow at this stage. If you turn `read_wal` into a table function in DuckDB you can produce DuckDB results - which can already be converted into whatever format you desire (pyarrow, Parquet, Pandas, ...). What you do with that result after in your own library is up to you obviously - but from DuckDB's side of things I think having a single table function called `read_wal` is sufficient to enable this use case.

Mytherin commented:
Just FYI - DuckDB does not always write records to the WAL. If we auto-checkpoint (which happens when doing bulk insertions) the WAL write is skipped to avoid doing a double write of large data sets. You will probably want to make this configurable for your use case - otherwise you will not see all data that ends up entering the system.

adsharma committed:
BufferedFileWriter does not lock by default

Caller will have to specify the precise locking they need.

On windows, mutual exclusion is enforced as follows:

GENERIC_WRITE, WRITE_LOCK | GENERIC_WRITE, WRITE_LOCK | one fails
GENERIC_WRITE, WRITE_LOCK | GENERIC_READ, READ_LOCK   | both pass
GENERIC_WRITE, WRITE_LOCK | GENERIC_READ, NO_LOCK     | both pass
GENERIC_READ, READ_LOCK   | GENERIC_READ, READ_LOCK   | both pass

PR #17423 for details on file locking behavior across platforms.

adsharma committed:
Add read_wal()

adsharma commented:
read_wal output looks like this:

```
┌─────────────────┬─────────────┬───────────────────────┬───────────────────────┐
│   record_type   │ record_size │      record_info      │      record_data      │
│     varchar     │    int64    │ map(varchar, varchar) │ map(varchar, varchar) │
├─────────────────┼─────────────┼───────────────────────┼───────────────────────┤
│ CREATE_INFO     │         141 │ {}                    │ {}                    │
│ FLUSH_INFO      │           0 │ {}                    │ {}                    │
│ TABLE_DATA_INFO │          20 │ {}                    │ {}                    │
│ TABLE_DATA_INFO │          93 │ {}                    │ {}                    │
│ FLUSH_INFO      │           0 │ {}                    │ {}                    │
│ TABLE_DATA_INFO │          20 │ {}                    │ {}                    │
│ TABLE_DATA_INFO │          93 │ {}                    │ {}                    │
│ FLUSH_INFO      │           0 │ {}                    │ {}                    │
│ TABLE_DATA_INFO │          20 │ {}                    │ {}                    │
│ TABLE_DATA_INFO │          93 │ {}                    │ {}                    │
├─────────────────┴─────────────┴───────────────────────┴───────────────────────┤
│ 10 rows                                                             4 columns │
└───────────────────────────────────────────────────────────────────────────────┘
```

adsharma commented on src/function/table/read_wal.cpp:
```diff
@@ -0,0 +1,112 @@
+#include "duckdb/main/client_context.hpp"
+#include "duckdb/storage/wal_reader.hpp"
+#include "duckdb/main/database_manager.hpp"
+//#include "json_serializer.hpp"
```
@Tishj @Mytherin it's not clear how to use the JSON serializer functionality here. It's in an extension.

Can we address that in another PR after this one is merged? If so, I'll remove the commented out code here and a few lines below.
